conda/core/solve.py

```diff
@@ -1272,27 +1273,83 @@ def _run_solver(self, state):
 
     def _build_transaction(self, state):
         from mamba import mamba_api as api
-        from mamba.utils import to_txn
+        from mamba.utils import to_package_record_from_subjson
+
+        solver = state["solver"]
+        index = state["index"]
+        installed_pkgs = state["installed_pkgs"]
 
         transaction = api.Transaction(
-            state["solver"],
+            solver,
             api.MultiPackageCache(context.pkgs_dirs),
             PackageCacheData.first_writable().pkgs_dir,
         )
         (names_to_add, names_to_remove), to_link, to_unlink = transaction.to_conda()
         specs_to_add = [MatchSpec(m) for m in names_to_add]
         specs_to_remove = [MatchSpec(m) for m in names_to_remove]
 
-        return to_txn(
-            specs_to_add,
-            specs_to_remove,
+        to_link_records, to_unlink_records = [], []
+
+        # What follows below is mainly mamba.utils.to_txn with some patches
+        # We will simplify this to save some transaction conversions
+        # whenever possible - TODO
+        conda_prefix_data = PrefixData(self.prefix)
+        final_precs = IndexedSet(conda_prefix_data.iter_records())
+
+        lookup_dict = {}
+        for _, entry in index:
+            lookup_dict[
+                entry["channel"].platform_url(entry["platform"], with_credentials=False)
+            ] = entry
+
+        for _, pkg in to_unlink:
+            for i_rec in installed_pkgs:
+                # Do not try to unlink virtual pkgs
+                if not i_rec.is_unmanageable and i_rec.fn == pkg:
+                    final_precs.remove(i_rec)
+                    to_unlink_records.append(i_rec)
+                    break
+            else:
+                print("No package record found!")
+
+        for c, pkg, jsn_s in to_link:
```

Comment:
What's `c` in this case?

Reply:
I want to think it's `c` for `channel` but I need to double check; this is coming straight from Mamba's code.

Reply:
Double checked and fixed.